harness/determined/core/_checkpoint.py

```diff
@@ -167,11 +393,107 @@ def store_path(
 
         self._report_checkpoint(storage_id, resources, metadata)
 
+    def _store_path_sharded(
+        self, metadata: Optional[Dict[str, Any]] = None
+    ) -> Iterator[Tuple[pathlib.Path, str]]:
+        storage_id = None
+        if self._dist.rank == 0:
+            storage_id = str(uuid.uuid4())
+        storage_id = self._dist.broadcast(storage_id)
+        assert storage_id
+
+        path = self._storage_manager.pre_store_path(storage_id)
+        yield path, storage_id
+
+        ckpt_dir = os.fspath(path)
+
+        if self._storage_manager.store_path_is_direct_access():
+            # Each rank saves files directly to ckpt_dir which means there is no conflict
+            # detection on upload. Metadata still needs to be merged and saved,
+            # and checkpoint has to be reported.
+            if self._dist.rank == 0:
+                resources = self._storage_manager._list_directory(ckpt_dir)
+
+            # Metadata should not be counted among resources. Why is that?
```

Comment:
Note, when you leave something like a `Why is that?` in a comment, please tag it with an `XXX` or something, to make it clear that you are effectively asking a question to the reviewer.

Don't let those sorts of things land in the codebase (the `XXX` makes it easy to detect unresolved tags)

Reply:
In this case, we've just always done it that way.  I'd be open to reversing that decision though if you'd like.

It used to make sense because we'd write metadata.json when we downloaded the checkpoint.  But we decided the checkpoint should have metadata all the time, even if we don't download it.  So I suppose it makes more sense now that metadata.json appears in the checkpoint manifest.

Reply:
Thanks for the explanation. I think we should be consistent between single and sharded upload, so that if single does not count metadata, then sharded also shouldn't. Let's keep it as is.